sourced/ml/algorithms/uast2nodes.py

```diff
@@ -0,0 +1,61 @@
+from collections import defaultdict
+
+from sourced.ml.transformers.quantization import Quantization
+
+
+class UastNode2Bag:
+    """
+    Converts a UAST to a bag of features that are node specific.
+    The features can be either (internal types, number of children)
+    """
+    def __init__(self, children2ind=None):
+        """
+        :param children2ind: The mapping from the number of children to bag of keys. \
+            :class:'Quantization' is used if it is not specified.
+        """
+        self.children2ind = Quantization() if children2ind is None else children2ind
+
+    def uast2nodes(self, root):
+        """
+        :param uast: The UAST root node.
+        :generate: The nodes composing the UAST.
+        """
+        queue = [root]
+        while queue:
+            child = queue.pop(0)
+            queue.extend(child.children)
+            yield child
+
+    def nodes2children(self, uast):
+        """
+        :param uast: The UAST root node.
+        :return: The list of the number of children of all these nodes.
+        """
+        children_freq = defaultdict(int)
+        for node in self.uast2nodes(uast):
+            children_freq[len(node.children)] += 1
+        return children_freq
+
+    def uast_to_bag(self, uast):
+        """
+        Converts a UAST to a bag of features. The weights are feature frequencies.
+        The number of children are preprocessed by :class:`Quantization`
+        :param uast: The UAST root node.
+        :return: bag of features
+        """
+        bag = defaultdict(int)
+        children_freq = self.nodes2children(uast)
+        self.children2ind.build_partition(children_freq)
```

Comment:
Moved from https://github.com/src-d/ml/pull/142#discussion_r157956161

`build_partition()` must be called global, not local. While building the bag, you need to emit two different kinds of features: children frequencies and the actual bags.

Reply:
Thus `InternalTypeChildrenBagExtractor` is not trivial: you need to override `BagsExtractor.inspect` to return both feature types. The core will do the dirty work for you and you will get the global children frequencies in `self.docfreq`

Reply:
Ok thanks, let me try this

Reply:
I don't understand the point in returning both children frequencies and the actual bag of features at the same time since the actual bags would be wrong because the number of children would not be quantized yet.
Or maybe you want me to process the quantization of children after the bags being built?

Reply:
Yes, after the bags are built. After the first stage, you get children freqs and individual role-children freqs. It contains everything you need to quantize children freqs and then merge several role-children together and assign the new counts